sklearn/preprocessing/_discretization.py

```diff
@@ -279,9 +292,9 @@ def transform(self, X):
         np.clip(Xt, 0, self.n_bins_ - 1, out=Xt)
 
         if self.encode == 'ordinal':
-            return Xt
+            return Xt.astype(self.dtype)
 
-        return self._encoder.transform(Xt)
+        return self._encoder.transform(Xt).astype(self.dtype)
```

Comment:
No instead, we want to pass the dtype parameter to `OneHotEncoder` in the `fit` method.